### PR TITLE
PS3 decrypted folder → ISO via makeps3iso (#98 Phase 2)

### PR DESCRIPTION
**Pull request by pacnpal**
## Summary

Adds the first **directory-as-input** conversion: a decrypted PS3 disc/JB folder → `.iso` using **makeps3iso** (`bucanero/ps3iso-utils`). This wires the Phase 1 directory seam (`InputKind`, `accepts_directory`, `tools_for_directory`, the `services.ps3` detector) end-to-end. Decryption, keys, pkg→folder, and disc→folder are all out of scope — this only repackages folders the user already decrypted.

### Decisions (confirmed before coding)
- **Provenance/license:** pin upstream `bucanero/ps3iso-utils` at a fixed commit; ship the unmodified **GPL-3.0** binary as a separate executable.
- **Verify:** no native verify → ship **no-verify / `supports_delete_on_verify=False`** plus a light, advisory **PARAM.SFO `TITLE_ID` readback** from the built ISO (never auto-deletes a curated source folder).
- **`-s` split flag:** always emit a single `.iso` (targets are ext4/NTFS/exFAT).
- **Listing cost:** always-on detector (cheap: `PS3_GAME/` stat + small SFO header read, off the event loop).

## What changed

- **Binary/Dockerfile:** pinned `makeps3iso` build stage mirroring maxcso; plain portable C (no x86 asm), so linux/amd64 + linux/arm64 both build. `makeps3iso_path` in `config.py`.
- **Service + tool:** `services/makeps3iso.py` (shell out via `SubprocessRunner.run`, normalized-basename output, partial-ISO cleanup on cancel/fail, TITLE_ID readback) and `MakePs3IsoTool` (one `folder_to_iso` `InputKind.DIRECTORY` mode, `accepts_directory` → `ps3.is_ps3_iso_source`, `detect_output`).
- **Pipeline:** `ConversionJob.input_kind` threaded end-to-end (`InputKind` relocated to `models` to avoid an import cycle, re-exported from `spec`); `plan_job` directory branch (validate `isdir` + detector, normalized basename, trailing-slash safe). The generic worker handles a directory source unchanged (archive-extract is gated on `::`).
- **Listing + search:** convertible directory rows annotated with `convertible_by` + sibling `.iso` in **both** `scan_directory` and the recursive search `_scan` (emitted as a job unit, not recursed past).
- **Lock manager:** a directory job protects its **whole subtree** — a per-file job / rename / delete inside an in-flight folder is rejected so it can't corrupt the ISO.
- **Frontend:** `folder_to_iso` tool/mode in `registry.js`; a *convertible* directory row becomes checkbox-selectable under a folder-input mode (name-click still navigates), while a plain directory stays navigation-only.
- **Docs:** `DESIGN_tool_plugin_architecture.md` §3.3.4 + `RELEASE_NOTES.md`.

## Testing

- `PYTHONPATH=app python -m pytest -q` — **994 passed, 1 skipped**
- `ruff check .` — clean; `pylint` on changed modules — 10.00/10
- `npm run build` — clean; ESLint clean; Svelte autofixer reports no issues
- New: `tests/test_makeps3iso.py` (detector→tool resolution, listing/search annotation, service convert argv + progress-to-100 + cancel-cleans-partial, `plan_job` valid/invalid/trailing-slash, lock-manager subtree protection); updated PS3 detector/registry/dispatch fixtures.

`hadolint` not available locally (CI runs it on release); the new stage mirrors the maxcso stage's `DL3008` ignore pattern.

## Out of scope / noted
- No key handling, no pkg→folder, no disc→folder, no decryption.
- A future folder→iso→chd would compose this directory input as step 1 of the existing Phase 1 `ChainSpec` — left as a noted possibility, not built.

https://claude.ai/code/session_018MavEx3D8vzgkj7t2ttsS5

---
_Generated by [Claude Code](https://claude.ai/code/session_018MavEx3D8vzgkj7t2ttsS5)_

## Summary by CodeRabbit

* **New Features**
  * PS3 folder → ISO conversion: repackage decrypted PS3 game/JB folders into mountable .iso via a new tool
  * File browser: convertible PS3 folders are selectable with annotated sibling .iso outputs
  * UI/tool registry: new PS3 ISO tool appears as a directory-only conversion mode

* **Improvements**
  * Progress streaming, cancellation handling, and automatic cleanup of partial ISOs
  * Directory-job subtree locking to prevent conflicting file operations
  * Post-build TITLE_ID readback reports mismatches without failing conversions

<details open><summary><h3>Greptile Summary</h3></summary>

This PR wires the first directory-as-input conversion end-to-end: a decrypted PS3 disc/JB folder → `.iso` via `makeps3iso`. It introduces `InputKind.DIRECTORY`, a new `MakePs3IsoTool` plugin, directory-aware lock manager subtree protection, a requeue/defer path for jobs blocked by an in-flight folder pack, and matching frontend selection/annotation for convertible folder rows.

- **Core conversion path**: `_plan_directory_job` validates source layout, guards against output-inside-source and out-of-volume edge cases, then routes through the standard `ConversionJob` lifecycle. `MakePs3IsoService.convert` runs `makeps3iso`, filters progress correctly, cleans partial ISOs on `BaseException`, and appends an advisory PARAM.SFO TITLE_ID readback as the terminal 100% update.
- **Lock management**: `acquire_dir_lock`/`release_dir_lock` protect the entire source subtree; `_defer_blocked_job` + `_schedule_dir_lock_requeue` implement wait-and-retry (not fail) for jobs that collide with an in-flight folder pack; `_requeue_tasks` now holds strong refs preventing premature GC of requeue tasks.
- **Frontend**: `allowsInputEntry` gates selection on `entry.type === 'directory'` + backend `convertible_by` annotation; the convertible hint badge now renders for both files and directories; `productPath` correctly derives the sibling `<folder>.iso` name.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge; the conversion pipeline, lock lifecycle, and partial-ISO cleanup are all correctly wired, and 994 tests pass.

The new code is carefully written with thorough edge-case handling: output-inside-source guard, volume boundary check, wait-and-retry for subtree lock collisions, and BaseException cleanup for partial ISOs. The two findings are both narrow defense-in-depth gaps that don't affect correctness in the common case.

app/services/job_manager.py (_directory_job_blocks pure-path vs realpath consistency) and app/services/makeps3iso.py (info() os.walk cost for large game folders).
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| app/services/makeps3iso.py | New service wrapping the makeps3iso binary; handles progress parsing, partial-ISO cleanup on BaseException, and advisory TITLE_ID readback. Clean design. |
| app/services/tools/makeps3iso.py | MakePs3IsoTool plugin: correctly declares DIRECTORY input_kind, guards detect_output with isdir, delegates all conversion to the service layer. |
| app/services/lock_manager.py | Adds acquire_dir_lock/release_dir_lock for subtree protection; _dir_overlaps_locked resolves symlinks via realpath, but _directory_job_blocks in job_manager uses a pure Path prefix check creating a minor symlink-bypass gap at the UI-layer check. |
| app/services/job_manager.py | Adds directory-job subtree awareness: _blocked_by_dir_lock, _defer_blocked_job, _schedule_dir_lock_requeue for wait-and-retry; input_kind threaded end-to-end; _requeue_tasks now holds strong refs preventing premature GC. |
| app/routes/convert.py | New _plan_directory_job validates isdir, source-layout detector, output-inside-source guard (realpath), and volume boundary; correctly plumbed through plan_job via input_kinds check. |
| app/routes/files.py | _detect_directory_outputs annotates convertible PS3 folders in both list_files and search_files; _assert_path_not_in_use extended with find_active_job_for_path for subtree containment checks. |
| Dockerfile | Adds makeps3iso-builder stage mirroring the maxcso pattern; pinned to an immutable commit hash for reproducible builds; GPL-3.0 binary shipped as standalone executable. |
| app/models.py | InputKind relocated here from spec.py to break the import cycle; ConversionJob gains input_kind field; Ps3IsoInfo model added. |
| src/lib/stores/conversion.svelte.js | Adds isDirectoryMode getter and allowsInputEntry() to gate selection on entry type + backend convertible_by annotation; clean entry-aware extension of allowsInput. |
| src/lib/tools/registry.js | Adds makeps3iso tool entry with inputKinds:['directory'], empty source/verify extensions, and correct productPath for folder→ISO naming. |

</details>

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant UI as Frontend
    participant Files as /api/files
    participant Convert as /api/convert
    participant JM as JobManager
    participant LM as LockManager
    participant Svc as MakePs3IsoService
    participant Bin as makeps3iso binary

    UI->>Files: GET /list (directory listing)
    Files->>Files: _detect_directory_outputs(item_path)
    Files->>Files: registry.tools_for_directory() → is_ps3_iso_source()
    Files-->>UI: "FileEntry{type:directory, convertible_by:['makeps3iso']}"

    UI->>Convert: "POST /convert {mode:folder_to_iso, file_path}"
    Convert->>Convert: plan_job → _plan_directory_job
    Convert->>Convert: Validate isdir + accepts_directory
    Convert->>Convert: Check output not inside source (realpath)
    Convert->>Convert: Check output within configured volumes
    Convert->>JM: "create_job(input_kind=DIRECTORY)"

    JM->>JM: _dispatch_job
    JM->>LM: is_within_locked_dir? (pre-check)
    LM-->>JM: false
    JM->>LM: acquire_lock(output.iso)
    LM-->>JM: acquired
    JM->>LM: acquire_dir_lock(source_folder)
    LM-->>JM: acquired (subtree locked)

    JM->>Svc: convert(folder, output.iso)
    Svc->>Bin: makeps3iso folder output.iso
    Bin-->>Svc: progress 0..99%
    Svc-->>JM: "yield {progress}"

    alt Cancellation / Error
        JM->>Svc: cancel
        Svc->>Svc: os.remove(output.iso) partial cleanup
    else Success
        Bin-->>Svc: exit 0
        Svc->>Svc: _readback_message TITLE_ID from ISO 9660
        Svc-->>JM: "yield {progress:100, message}"
    end

    JM->>LM: release_lock(output.iso)
    JM->>LM: release_dir_lock(source_folder)

    Note over JM,LM: Concurrent file job targeting path inside source_folder
    JM->>LM: acquire_lock(inner_file)
    LM-->>JM: false within_locked_dir
    JM->>JM: _defer_blocked_job → sleep 2s → requeue
```
</details>

<details open><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `app/services/lock_manager.py`, line 732-746 ([link](https://github.com/pacnpal/compressatorium/blob/07cb69017c23c10b69dac95088b569f1d0205892/app/services/lock_manager.py#L732-L746)) 

   <a href="#">P2</a> **Disk I/O (`os.path.realpath`) while holding `_lock_mutex`**

   `_dir_overlaps_locked` iterates over `self._locks` and calls `self._resolve(p)` (i.e. `os.path.realpath`) on each currently-held file-lock path. This is disk I/O executed under `_lock_mutex`. For a large batch job with many simultaneous file locks, every call to `acquire_dir_lock` or `dir_lock_would_conflict` holds the mutex while resolving N symlinks — blocking other threads that need the lock (e.g. concurrent `acquire_lock` calls) for the full duration.

   Consider pre-resolving and caching the resolved form of each file-lock path when it is first acquired (already done for dir-lock entries, which are stored as their resolved path) so `_dir_overlaps_locked` can do a pure in-memory set lookup instead.

   <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20app%2Fservices%2Flock_manager.py%0ALine%3A%20732-746%0A%0AComment%3A%0A**Disk%20I%2FO%20%28%60os.path.realpath%60%29%20while%20holding%20%60_lock_mutex%60**%0A%0A%60_dir_overlaps_locked%60%20iterates%20over%20%60self._locks%60%20and%20calls%20%60self._resolve%28p%29%60%20%28i.e.%20%60os.path.realpath%60%29%20on%20each%20currently-held%20file-lock%20path.%20This%20is%20disk%20I%2FO%20executed%20under%20%60_lock_mutex%60.%20For%20a%20large%20batch%20job%20with%20many%20simultaneous%20file%20locks%2C%20every%20call%20to%20%60acquire_dir_lock%60%20or%20%60dir_lock_would_conflict%60%20holds%20the%20mutex%20while%20resolving%20N%20symlinks%20%E2%80%94%20blocking%20other%20threads%20that%20need%20the%20lock%20%28e.g.%20concurrent%20%60acquire_lock%60%20calls%29%20for%20the%20full%20duration.%0A%0AConsider%20pre-resolving%20and%20caching%20the%20resolved%20form%20of%20each%20file-lock%20path%20when%20it%20is%20first%20acquired%20%28already%20done%20for%20dir-lock%20entries%2C%20which%20are%20stored%20as%20their%20resolved%20path%29%20so%20%60_dir_overlaps_locked%60%20can%20do%20a%20pure%20in-memory%20set%20lookup%20instead.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=pacnpal%2Fcompressatorium&pr=158&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22pacnpal%2Fcompressatorium%22%20on%20the%20existing%20branch%20%22claude%2Fcool-noether-7lo4ey%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22claude%2Fcool-noether-7lo4ey%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20app%2Fservices%2Flock_manager.py%0ALine%3A%20732-746%0A%0AComment%3A%0A**Disk%20I%2FO%20%28%60os.path.realpath%60%29%20while%20holding%20%60_lock_mutex%60**%0A%0A%60_dir_overlaps_locked%60%20iterates%20over%20%60self._locks%60%20and%20calls%20%60self._resolve%28p%29%60%20%28i.e.%20%60os.path.realpath%60%29%20on%20each%20currently-held%20file-lock%20path.%20This%20is%20disk%20I%2FO%20executed%20under%20%60_lock_mutex%60.%20For%20a%20large%20batch%20job%20with%20many%20simultaneous%20file%20locks%2C%20every%20call%20to%20%60acquire_dir_lock%60%20or%20%60dir_lock_would_conflict%60%20holds%20the%20mutex%20while%20resolving%20N%20symlinks%20%E2%80%94%20blocking%20other%20threads%20that%20need%20the%20lock%20%28e.g.%20concurrent%20%60acquire_lock%60%20calls%29%20for%20the%20full%20duration.%0A%0AConsider%20pre-resolving%20and%20caching%20the%20resolved%20form%20of%20each%20file-lock%20path%20when%20it%20is%20first%20acquired%20%28already%20done%20for%20dir-lock%20entries%2C%20which%20are%20stored%20as%20their%20resolved%20path%29%20so%20%60_dir_overlaps_locked%60%20can%20do%20a%20pure%20in-memory%20set%20lookup%20instead.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=pacnpal%2Fcompressatorium&pr=158&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a>
</details>

<sub>Reviews (2): Last reviewed commit: ["job\_manager: keep strong refs to in-flig..."](https://github.com/pacnpal/compressatorium/commit/96c44dc3c3eb9db48c8d7c9b165ceb8ffff627a6) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=37266235)</sub>

### CODE REVIEW

File: app/services/job_manager.py [hunk 1]
```diff
@@ -596,6 +642,10 @@ def _is_path_in_use_by_other_job(self, job_id: str, path: str) -> bool:
                 continue
             if job.mode in _EXTERNAL_JOB_MODES:
                 continue
+            # Reject a delete that targets a path inside another active
+            # directory job's source folder (its whole subtree is in use).
+            if self._directory_job_blocks(job, target):
+                return True
```
[the remark this answers is not in the document]
Comment (pacnpal) [reply app/services/job_manager.py:723]: Thanks — I'm leaving this one as-is for this PR. The `_is_path_in_use_by_other_job` call at this site is pre-existing code in the delete-on-verify safety loop and isn't exercised by the makeps3iso feature: `folder_to_iso` ships with `supports_delete_on_verify=False`, so a directory job never enters this branch. `_normalize_path` resolves already-validated local paths (no network FS assumed), so the per-path cost is small, and offloading this security-sensitive delete loop to a threadpool is a broader change I'd rather not fold into this feature PR.

The new directory code path this PR *does* add — the descendant-of-active-directory-job check used by rename/delete — already runs off the event loop: `files.py:_assert_path_not_in_use` calls `find_active_job_for_path` via `run_in_threadpool`. Happy to open a follow-up if you'd like the pre-existing delete-on-verify loop offloaded separately.

---
_Generated by [Claude Code](https://claude.ai/code)_
